src/main/java/org/semux/core/TransactionExecutor.java

```diff
@@ -180,33 +193,82 @@ public TransactionExecutor(Config config) {
                 }
                 break;
             }
+
+            case CALL:
+            case CREATE:
+                long maxGasFee = tx.getGas() * tx.getGasPrice();
+                Amount maxCost = sum(sum(value, fee), Unit.NANO_SEM.of(maxGasFee));
+                if (maxCost.lte(available)) {
+                    // VM calls still use fees
+                    as.adjustAvailable(from, neg(sum(value, fee)));
+
+                    if (tx.getGas() > config.vmMaxBlockGasLimit()) {
+                        result.setCode(Code.INVALID_GAS);
+                    } else if (block == null) {
+                        // workaround for pending manager so it doesn't execute these
+                        // we charge gas later
+                        as.increaseNonce(from);
+                        result.setCode(Code.SUCCESS);
+                    } else {
+                        executeVmTransaction(result, tx, as, block, gasUsedInBlock);
+                        gasUsedInBlock += result.getGasUsed();
+                    }
+                } else {
+                    result.setCode(Code.INSUFFICIENT_AVAILABLE);
+                }
+                break;
+
             default:
                 // unsupported transaction type
                 result.setCode(Code.INVALID_TYPE);
                 break;
             }
 
-            // increase nonce if success or failure
-            if (result.getCode().isAccepted()) {
+            // increase nonce if success
+            // creates and calls increase their own nonces internal to VM
+            if (result.getCode().isAccepted() && type != TransactionType.CREATE && type != TransactionType.CALL) {
                 as.increaseNonce(from);
             }
         }
 
         return results;
     }
 
+    private void executeVmTransaction(TransactionResult result, Transaction tx, AccountState as, SemuxBlock block,
+            long gasUsedInBlock) {
+        SemuxTransaction transaction = new SemuxTransaction(tx);
+        Repository repository = new SemuxRepository(as);
+        ProgramInvokeFactory invokeFactory = new ProgramInvokeFactoryImpl();
+
+        org.ethereum.vm.client.TransactionExecutor executor = new org.ethereum.vm.client.TransactionExecutor(
+                transaction, block, repository, blockStore,
+                Spec.DEFAULT, invokeFactory, gasUsedInBlock, false);
+
+        TransactionReceipt summary = executor.run();
+        if (summary == null) {
+            result.setCode(Code.SUCCESS);
+        } else {
+            for (LogInfo log : summary.getLogs()) {
+                result.addLog(log);
+            }
+            result.setGasUsed(summary.getGasUsed());
```

Comment:
Need to add integration tests about `REVERT` and `INVALID` for the `getGasUsed()`.